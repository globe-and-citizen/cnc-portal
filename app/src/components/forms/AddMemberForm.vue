<template>
  <h1 class="font-bold text-2xl">Add New Member</h1>
  <hr class="" />
  <div v-for="(input, index) in formData" :key="index" class="input-group mt-3">
    <label
      class="input input-bordered flex items-center gap-2 input-md"
      :class="{ 'input-error': !input.isValid }"
    >
      <input
        type="text"
        class="w-24"
        v-model="input.name"
        @keyup.stop="
          () => {
            emits('searchUsers', input)
            dropdown = true
          }
        "
        :placeholder="'Member Name ' + (index + 1)"
      />
      |
      <input
        type="text"
        class="grow"
        v-model="input.address"
        @input="() => (isAddress(input.address) ? (input.isValid = true) : (input.isValid = false))"
        @keyup.stop="
          () => {
            emits('searchUsers', input)
            dropdown = true
          }
        "
        :placeholder="'Wallet Address ' + (index + 1)"
      />
      <span class="badge badge-primary">Mandatory</span>
    </label>
  </div>
  <div class="dropdown" :class="{ 'dropdown-open': !!users && users.length > 0 }" v-if="dropdown">
    <ul class="p-2 shadow menu dropdown-content z-[1] bg-base-100 rounded-box w-96">
      <li v-for="user in users" :key="user.address">
        <a
          @click="
            () => {
              const l = formData.length - 1
              if (l >= 0) {
                formData[l].name = user.name ?? ''
                formData[l].address = user.address ?? ''
                if (isAddress(formData[l].address)) {
                  formData[l].isValid = true
                } else {
                  formData[l].isValid = false
                }
                dropdown = false
              }
            }
          "
        >
          {{ user.name }} | {{ user.address }}
        </a>
      </li>
    </ul>
  </div>
  <div class="flex justify-end pt-3">
    <div
      class="w-6 h-6 cursor-pointer"
      @click="
        () => {
          formData.push({ name: '', address: '', isValid: false })
        }
      "
      data-test="plus-icon"
    >
      <PlusCircleIcon class="size-6 text-green-700" />
    </div>
    <div
      class="w-6 h-6 cursor-pointer"
      @click="
        () => {
          if (formData.length > 1) {
            formData.pop()
          }
        }
      "
      data-test="minus-icon"
    >
      <MinusCircleIcon class="size-6 text-red-700" />
    </div>
  </div>
  <div class="flex justify-center">
    <LoadingButton color=" primary w-24" v-if="isLoading" class="btn btn-primary justify-center" />
    <button class="btn btn-primary justify-center" v-else @click="emits('addMembers')">Add</button>
  </div>
</template>
<script setup lang="ts">
import { ref } from 'vue'
import type { User } from '@/types'
// import { PlusCircleIcon, MinusCircleIcon } from '@heroicons/vue/24/outline'
import { PlusCircleIcon, MinusCircleIcon } from '@heroicons/vue/24/solid'
import LoadingButton from '../LoadingButton.vue'
import { isAddress } from 'ethers'
const emits = defineEmits(['updateForm', 'addMembers', 'searchUsers'])
const props = defineProps<{
  formData: Array<{ name: string; address: string; isValid: boolean }>
  users: User[]
  isLoading: boolean
}>()
const dropdown = ref<boolean>(true)

const formData = ref(props.formData)
</script>
