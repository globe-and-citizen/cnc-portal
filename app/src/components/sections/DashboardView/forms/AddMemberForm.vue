<template>
  <h1 class="font-bold text-2xl">Add New Member</h1>
  <hr />
  <div class="flex flex-col gap-5 pt-5">
    <MultiSelectMemberInput v-model="formData" />

    <div v-if="addMembersError">
      <div class="alert alert-warning" v-if="statusCode === 401">
        You don't have the right for this
      </div>
      <div class="alert alert-danger" v-else>Something went wrong, Unable to add team Members</div>
    </div>

    <ButtonUI
      variant="primary"
      class="justify-center"
      :loading="addMembersLoading"
      :disabled="addMembersLoading"
      @click="
        async () => {
          await executeAddMembers()
          teamStore.fetchTeam(String(props.teamId))
        }
      "
      >Add Members</ButtonUI
    >
  </div>

  <div class="divider m-0"></div>
</template>
<script setup lang="ts">
import { computed, ref, watch } from 'vue'
import ButtonUI from '@/components/ButtonUI.vue'
import MultiSelectMemberInput from '@/components/utils/MultiSelectMemberInput.vue'
import { useCustomFetch } from '@/composables/useCustomFetch'
import { useToastStore, useTeamStore } from '@/stores'
import type { User } from '@/types'

const emits = defineEmits(['memberAdded'])
const { addSuccessToast, addErrorToast } = useToastStore()
const teamStore = useTeamStore()

const props = defineProps<{
  teamId: string | number
}>()

const formData = ref<Array<Pick<User, 'address' | 'name'>>>([])

const membersAddress = computed(() =>
  formData.value.map((member) => {
    return { address: member.address }
  })
)
const {
  execute: executeAddMembers,
  error: addMembersError,
  isFetching: addMembersLoading,
  statusCode
} = useCustomFetch(`teams/${props.teamId}/member`, { immediate: false })
  .post(membersAddress)
  .json<{ member: string }>()

watch(addMembersError, () => {
  if (addMembersError.value) {
    addErrorToast(addMembersError.value || 'Failed to add members')
  }
})

watch(addMembersLoading, () => {
  if (!addMembersLoading.value && !addMembersError.value) {
    addSuccessToast('Members added successfully')
    emits('memberAdded')
  }
})
</script>
