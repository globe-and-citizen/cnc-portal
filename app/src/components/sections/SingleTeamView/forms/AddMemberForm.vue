<template>
  <h1 class="font-bold text-2xl">Add New Member</h1>
  <hr />
  <div v-for="(input, index) in formData" :key="index" class="input-group mt-3">
    <label class="input input-bordered flex items-center gap-2 input-md">
      <input
        type="text"
        class="w-24"
        v-model="input.name"
        @keyup.stop="
          () => {
            emits('searchUsers', input)
            dropdown = true
          }
        "
        :placeholder="'Member Name ' + (index + 1)"
      />
      |
      <input
        type="text"
        class="grow"
        v-model="input.address"
        @keyup.stop="
          () => {
            emits('searchUsers', input)
            dropdown = true
          }
        "
        :placeholder="'Wallet Address ' + (index + 1)"
      />
      <span class="badge badge-primary">Mandatory</span>
    </label>
  </div>

  <div class="dropdown" :class="{ 'dropdown-open': !!users && users.length > 0 }" v-if="dropdown">
    <ul class="p-2 shadow menu dropdown-content z-[1] bg-base-100 rounded-box w-96">
      <li v-for="user in users" :key="user.address">
        <a
          @click="
            () => {
              const l = formData.length - 1
              if (l >= 0) {
                formData[l].name = user.name ?? ''
                formData[l].address = user.address ?? ''
                dropdown = false
              }
            }
          "
        >
          {{ user.name }} | {{ user.address }}
        </a>
      </li>
    </ul>
  </div>
  <div
    class="pl-4 text-red-500 text-sm w-full text-left"
    v-for="error of $v.formData.$errors"
    :key="error.$uid"
  >
    <div v-if="Array.isArray(error.$message)">
      <div v-for="(errorObj, index) of error.$message" :key="index">
        <div v-for="(error, index1) of errorObj" :key="index1">
          {{ error }}
        </div>
      </div>
    </div>
    <div v-else>
      {{ error.$message }}
    </div>
  </div>
  <div class="flex justify-end pt-3">
    <div
      class="w-6 h-6 cursor-pointer"
      @click="() => formData.push({ name: '', address: '' })"
      data-test="plus-icon"
    >
      <PlusCircleIcon class="size-6 text-green-700" />
    </div>
    <div
      class="w-6 h-6 cursor-pointer"
      @click="() => formData.length > 1 && formData.pop()"
      data-test="minus-icon"
    >
      <MinusCircleIcon class="size-6 text-red-700" />
    </div>
  </div>
  <div class="flex justify-center">
    <LoadingButton color="primary w-24" v-if="isLoading" class="btn btn-primary justify-center" />
    <button class="btn btn-primary justify-center" v-else @click="submitForm">Add</button>
  </div>
</template>
<script setup lang="ts">
import { ref } from 'vue'
import { useVuelidate } from '@vuelidate/core'
import { required, helpers } from '@vuelidate/validators'
import type { User } from '@/types'
import { PlusCircleIcon, MinusCircleIcon } from '@heroicons/vue/24/solid'
import LoadingButton from '@/components/LoadingButton.vue'
import { isAddress } from 'ethers'

const emits = defineEmits(['updateForm', 'addMembers', 'searchUsers'])

const props = defineProps<{
  formData: Array<{ name: string; address: string }>
  users: User[]
  isLoading: boolean
}>()

const formData = ref(props.formData)
const dropdown = ref<boolean>(true)

const rules = {
  formData: {
    $each: helpers.forEach({
      address: {
        required: helpers.withMessage('Address is required', required),
        $valid: helpers.withMessage('Invalid wallet address', (value: string) => isAddress(value))
      }
    }),

    $valid: helpers.withMessage(
      'At least one member is required',
      (value: Array<{ name: string; address: string }>) => {
        return value.some((v) => v.name && v.address)
      }
    )
  }
}

const $v = useVuelidate(rules, { formData })

const submitForm = () => {
  $v.value.$touch()
  if ($v.value.$invalid) return
  emits('addMembers', formData.value)
}
</script>
