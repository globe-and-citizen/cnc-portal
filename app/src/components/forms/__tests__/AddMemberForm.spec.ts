import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import AddMemberForm from '@/components/forms/AddMemberForm.vue'

describe('AddMemberModal.vue', () => {
  const formData = [
    { name: 'Hermann', address: '0xaFeF48F7718c51fb7C6d1B314B3991D2e1d8421E', isValid: true }
  ]
  const users = [
    { name: 'Dachu', address: '0xaFeF48F7718c51fb7C6d1B314B3991D2e1d8421E' },
    { name: 'Farrell', address: '0x4b6Bf5cD91446408290725879F5666dcd9785F62' }
  ]

  const wrapper = mount(AddMemberForm, {
    props: {
      formData,
      users,
      isLoading: false,
      showAddMemberForm: true
    }
  })
  // Test the rendering of the component
  describe('Render', () => {
    it('renders correctly with initial props', () => {
      expect(wrapper.find('h1').text()).toBe('Add New Member')
      expect(wrapper.findAll('.input-group').length).toBe(formData.length)
    })
    it('shows dropdown when users are available', async () => {
      expect(wrapper.find('.dropdown-open').exists()).toBe(true)
    })

    it('updates formData when a user is selected from dropdown', async () => {
      await wrapper.find('.dropdown a').trigger('click')
      expect((wrapper.vm as any).formData[0].name).toBe(users[0].name)
      expect((wrapper.vm as any).formData[0].address).toBe(users[0].address)
    })
  })

  // Test the emitting of events
  describe('Emits', () => {
    it('emits addMembers when add button is clicked', async () => {
      await wrapper.find('button.btn-primary').trigger('click')
      expect(wrapper.emitted('addMembers')).toBeTruthy()
    })

    it('Should trigger searchUsers on keyup.stop', async () => {
      const wrapper = mount(AddMemberForm, {
        props: {
          formData: [{ name: '', address: '', isValid: false }],
          users,
          isLoading: false
        }
      })
      const inputs = wrapper.findAll('input')
      expect(inputs.length).toBe(2)

      // Intial state of the searchUser event
      expect(wrapper.emitted('searchUsers')).toBe(undefined)

      // Set the value of the first input and trigger
      await inputs[0].setValue('Farrel')

      // No emit after setting the value until we trigger
      expect(wrapper.emitted('searchUsers')).toBe(undefined)

      expect(inputs[0].element.value).toBe('Farrel')
      await inputs[0].trigger('keyup.stop')

      // Result after trigger
      expect(wrapper.emitted('searchUsers')).toMatchSnapshot(`
        [
          [
            {
              "address": "",
              "isValid": false,
              "name": "Farrel",
            },
          ],
        ]
      `)

      // Type in the second input and trigger
      await inputs[1].setValue('0xaFeF48F7718c51fb7C6d1B314B3991D2e1d8421E')
      await inputs[1].trigger('keyup.stop')

      // Result After trigger
      expect(wrapper.emitted('searchUsers')).toMatchSnapshot(`
        [
          [
            {
              "address": "0xaFeF48F7718c51fb7C6d1B314B3991D2e1d8421E",
              "isValid": true,
              "name": "Farrel",
            },
          ],
          [
            {
              "address": "0xaFeF48F7718c51fb7C6d1B314B3991D2e1d8421E",
              "isValid": true,
              "name": "Farrel",
            },
          ],
        ]
      `)
    })
  })

  // The the behavior of the component on user actions
  describe('Actions', () => {
    it('adds a new member input field when clicking the add icon', async () => {
      const wrapper = mount(AddMemberForm, {
        props: {
          formData: [{ name: '', address: '', isValid: false }],
          users: [],
          isLoading: false
        }
      })

      const addButton = wrapper.find('[data-test="plus-icon"]')
      await addButton.trigger('click')

      expect(wrapper.findAll('.input-group').length).toBe(2)
    })

    it('removes the last member input field when clicking the remove icon', async () => {
      const wrapper = mount(AddMemberForm, {
        props: {
          formData: [
            { name: '', address: '', isValid: false },
            { name: '', address: '', isValid: false }
          ],
          users: [],
          isLoading: false
        }
      })
      const removeButton = wrapper.find('[data-test="minus-icon"]')
      await removeButton.trigger('click')

      expect(wrapper.findAll('.input-group').length).toBe(1)
    })
  })
})
