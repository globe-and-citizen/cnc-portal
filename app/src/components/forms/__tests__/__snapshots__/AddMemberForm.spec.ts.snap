// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddMemberModal.vue > Emits > Should trigger searchUsers on keyup.stop > 
        [
          [
            {
              "address": "",
              "isValid": false,
              "name": "Farrel",
            },
          ],
        ]
       1`] = `
[
  [
    {
      "address": "",
      "isValid": false,
      "name": "Farrel",
    },
  ],
]
`;

exports[`AddMemberModal.vue > Emits > Should trigger searchUsers on keyup.stop > 
        [
          [
            {
              "address": "0xaFeF48F7718c51fb7C6d1B314B3991D2e1d8421E",
              "isValid": true,
              "name": "Farrel",
            },
          ],
          [
            {
              "address": "0xaFeF48F7718c51fb7C6d1B314B3991D2e1d8421E",
              "isValid": true,
              "name": "Farrel",
            },
          ],
        ]
       1`] = `
[
  [
    {
      "address": "0xaFeF48F7718c51fb7C6d1B314B3991D2e1d8421E",
      "isValid": true,
      "name": "Farrel",
    },
  ],
  [
    {
      "address": "0xaFeF48F7718c51fb7C6d1B314B3991D2e1d8421E",
      "isValid": true,
      "name": "Farrel",
    },
  ],
]
`;

exports[`AddMemberModal.vue > Snapshot > matches the snapshot 1`] = `
"<h1 class="font-bold text-2xl">Add New Member</h1>
<hr>
<div class="input-group mt-3"><label class="input input-bordered flex items-center gap-2 input-md"><input type="text" class="w-24" placeholder="Member Name 1"> | <input type="text" class="grow" placeholder="Wallet Address 1"><span class="badge badge-primary">Mandatory</span></label></div>
<!--v-if-->
<div class="flex justify-end pt-3">
  <div class="w-6 h-6 cursor-pointer" data-test="plus-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="currentColor" aria-hidden="true" data-slot="icon" class="size-6 text-green-700">
      <path fill-rule="evenodd" d="M12 2.25c-5.385 0-9.75 4.365-9.75 9.75s4.365 9.75 9.75 9.75 9.75-4.365 9.75-9.75S17.385 2.25 12 2.25ZM12.75 9a.75.75 0 0 0-1.5 0v2.25H9a.75.75 0 0 0 0 1.5h2.25V15a.75.75 0 0 0 1.5 0v-2.25H15a.75.75 0 0 0 0-1.5h-2.25V9Z" clip-rule="evenodd"></path>
    </svg></div>
  <div class="w-6 h-6 cursor-pointer" data-test="minus-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="currentColor" aria-hidden="true" data-slot="icon" class="size-6 text-red-700">
      <path fill-rule="evenodd" d="M12 2.25c-5.385 0-9.75 4.365-9.75 9.75s4.365 9.75 9.75 9.75 9.75-4.365 9.75-9.75S17.385 2.25 12 2.25Zm3 10.5a.75.75 0 0 0 0-1.5H9a.75.75 0 0 0 0 1.5h6Z" clip-rule="evenodd"></path>
    </svg></div>
</div>
<div class="flex justify-center"><button class="btn btn-primary justify-center">Add</button></div>"
`;
